### DISCUSSION TITLE
ANSIBLE_ROLES_PATH and ANSIBLE_LIBRARY do not respect the Ansible defaults

### DISCUSSION THREAD

**Discussion by joren485** [ansible/ansible-lint#1353] [Q&A] [answered]
##### Summary

Ansible Lint uses the environment variables `ANSIBLE_LIBRARY` and `ANSIBLE_ROLES_PATH` to configure custom module and role paths, respectively. These environment variables seem to be taken from [Ansible](https://docs.ansible.com/ansible/latest/reference_appendices/config.html). However, in Ansible the configuration options that these environment variables provide have defaults. Given that these variables are taken from Ansible and they are not explicitly documented by Ansible Lint, it would be reasonable to assume the same defaults are used in Ansible Lint. However, Ansible Lint does not do this.
 
On a related note, these options can also be configured in `ansible.cfg` in Ansible. Ansible Lint does not seem to use `ansible.cfg`, adding to the confusion.

##### Issue Type

- Bug Report

**Comment by ssbarnea** [comment]
I converted this into a discussion because is not really a bug, at least not yet.  Ansible lint does not care about ansible.cfg and it should never do, because that's Ansible job to look at.

When Ansible runs, it loads both config file and environment variables and it is well documented that presence of environment variables does override values from config file. If you define these paths inside your ansible.cfg file it would be up to you to find a solution, likely you will have to also define same value as environment variables, so Ansible can still make use of them.

Keep in mind that this issue mostly unrelated to the linter as if you manage to make `ansible-playbook --syntax-check` to work, it will likely work with the linter too.

Our linter injects additional temp directories to some of these variables by appending to them and does this only when is needed: requirements.yml exists or mocking options are used. It does try to help ansible find the potentially missing bits in a way that is least intrusive for the Ansible execution environment.

For the moment I do not see any bug here and clear the report is incomplete as it does not even mention version of Ansible used, linter or pointing to a basic repository where the issue can be reproduced. I am sure that there are still bugs around but we need more info to identify them.

I would be more than happy to welcome improvement pull-requests, even for docs as I know that the injection of those variables is not really documented, we only display what we do but not really documenting why.

**Reply by joren485** [reply]
Thanks for the detailed explanation. That clears up a lot for me.

What might be considered a bug is that referenced roles that are defined in one of the default paths ([`~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles`](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-roles-path)) are not found by Ansible Lint (but are by Ansible and `ansible-playbook --syntax-check`).

For example if we have the following two files:
`test-playbook.yaml`:
```
- name: "Test Playbook"
  hosts: "localhost"
  roles:
    - "test-role"
```

`~/.ansible/roles/test-role/tasks/main.yaml`:
```
- debug:
    msg: "Test Message"
```
Running `test-playbook.yaml` with `ansible-playbook` works and `ansible-playbook test-playbook.yaml --syntax-check` works. However, `ansible-lint` will error `test-playbook.yaml:4:7 ERROR! the role 'test-role' was not found`.

Would you consider this intended behaviour or a bug?

**Reply by joren485** [reply]
It does work if `test-role` is added to `mock_roles`. However the documentation says that `mock_roles` are "roles in order to pass ansible-playbook --syntax-check".

**Comment by ssbarnea** [comment, marked as answer]
If ansible syntax check is passing outside linter but not inside it, is a linter bug we need to fix.

**Reply by joren485** [reply]
I have opened #1356 to give a detailed explanation of this issue.
